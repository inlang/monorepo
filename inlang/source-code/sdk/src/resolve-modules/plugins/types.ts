import type { LanguageTag } from "@inlang/language-tag"
import type { NodeishFilesystem as LisaNodeishFilesystem } from "@lix-js/fs"
import type {
	PluginReturnedInvalidCustomApiError,
	PluginLoadMessagesFunctionAlreadyDefinedError,
	PluginSaveMessagesFunctionAlreadyDefinedError,
	PluginHasInvalidIdError,
	PluginHasInvalidSchemaError,
	PluginUsesReservedNamespaceError,
	PluginsDoNotProvideLoadOrSaveMessagesError,
} from "./errors.js"
import type { Message } from "@inlang/message"
import type { JSONObject } from "@inlang/json-types"
import type { CustomApiInlangIdeExtension, Plugin } from "@inlang/plugin"

/**
 * The filesystem is a subset of project lisa's nodeish filesystem.
 *
 * - only uses minimally required functions to decrease the API footprint on the ecosystem.
 */
export type NodeishFilesystemSubset = Pick<
	LisaNodeishFilesystem,
	"readFile" | "readdir" | "mkdir" | "writeFile"
>

/**
 * Function that resolves (imports and initializes) the plugins.
 */
export type ResolvePluginsFunction = (args: {
	plugins: Array<Plugin>
	settings: Record<Plugin["id"], JSONObject>
	nodeishFs: NodeishFilesystemSubset
}) => Promise<{
	data: ResolvedPluginApi
	errors: Array<
		| PluginReturnedInvalidCustomApiError
		| PluginLoadMessagesFunctionAlreadyDefinedError
		| PluginSaveMessagesFunctionAlreadyDefinedError
		| PluginHasInvalidIdError
		| PluginHasInvalidSchemaError
		| PluginUsesReservedNamespaceError
		| PluginsDoNotProvideLoadOrSaveMessagesError
	>
}>

/**
 * The API after resolving the plugins.
 */
export type ResolvedPluginApi = {
	loadMessages: (args: {
		languageTags: LanguageTag[]
		sourceLanguageTag: LanguageTag
	}) => Promise<Message[]> | Message[]
	saveMessages: (args: { messages: Message[] }) => Promise<void> | void
	/**
	 * App specific APIs.
	 *
	 * @example
	 *  // define
	 *  customApi: ({ settings }) => ({
	 * 	 "app.inlang.ide-extension": {
	 * 	   messageReferenceMatcher: () => {
	 * 		 // use settings
	 * 		 settings.pathPattern
	 * 		return
	 * 	   }
	 * 	 }
	 *  })
	 *  // use
	 *  customApi['app.inlang.ide-extension'].messageReferenceMatcher()
	 */
	customApi: Record<`app.${string}.${string}` | `library.${string}.${string}`, unknown> & {
		"app.inlang.ideExtension"?: CustomApiInlangIdeExtension
	}
}
