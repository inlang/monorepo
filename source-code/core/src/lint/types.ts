export type MaybePromise<T> = T | Promise<T>

export type Pretty<T> = T extends (...args: any[]) => any
	? T
	: T extends abstract new (...args: any[]) => any
	? T
	: { [K in keyof T]: T[K] }
