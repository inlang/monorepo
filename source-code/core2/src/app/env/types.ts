/**
 * The inlang environment.
 *
 * Read more https://inlang.com/documentation/inlang-environment
 */
export type InlangInstanceEnv = {
	fs: any
	import: any
}
