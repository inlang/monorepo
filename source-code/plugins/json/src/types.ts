export type SerializedMessage = {
	id: string
	text: string
} & MessageMetadata

export type MessageMetadata = {
	parentKeys?: string[]
	fileName?: string
	keyName?: string
}
