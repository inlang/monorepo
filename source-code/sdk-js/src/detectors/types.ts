import type { LanguageTag } from '@inlang/core/languageTag'

type MaybePromise<T> = T | Promise<T>

export type Detector<Parameters extends unknown[] = never[]> = (
	...parameters: Parameters
) => MaybePromise<LanguageTag[]>

export type DetectorInitializer<Parameters extends unknown[] = never[]> = (
	...parameters: Parameters
) => Detector<Parameters>
