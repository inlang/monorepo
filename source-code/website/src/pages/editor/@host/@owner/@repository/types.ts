/** @example `{host}/{owner}/{repository}` */
export type EditorRouteParams = {
	/** @example `github.com` */
	host: string
	/** @example `inlang` */
	owner: string
	/** @example `website` */
	repository: string
}

export type EditorSearchParams = {
	/**
	 * the current branch
	 *
	 * not implemented yet see https://github.com/inlang/inlang/discussions/166
	 */
	// branch?: string;
}
