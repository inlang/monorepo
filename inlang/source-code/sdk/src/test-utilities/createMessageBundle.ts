import * as MF from "messageformat"
import type * as AST from "../v2/types.js"

/**
 * Utility function to generate message bundles for tests
 *
 * @throws if the given source-strings aren't valid or not representable
 */
export function createMessageBundle(bundleOptions: {
	id: string
	aliases?: Record<string, string>
	/**
	 * A map of languages to ICU MessageFormat 2 message strings
	 */
	messages: Record<string, string>
}): AST.MessageBundle {
	const bundle: AST.MessageBundle = {
		id: bundleOptions.id,
		alias: bundleOptions.aliases || {},

		messages: Object.entries(bundleOptions.messages).map(([locale, source]) =>
			parseICUMessage(locale, source)
		),
	}

	return bundle
}

function parseICUMessage(locale: string, source: string): AST.Message {
	const ast = MF.parseMessage(source)
	const declarations = ast.declarations.map(toDeclaration)

	switch (ast.type) {
		case "message": {
			return {
				locale,
				declarations,
				selectors: [],
				variants: [
					{
						match: [],
						pattern: toPattern(ast.pattern),
					},
				],
			}
		}
		case "select": {
			return {
				locale,
				declarations,
				selectors: ast.selectors.map(toExpression),
				variants: ast.variants.map(toVariant),
			}
		}
	}
}

function toVariant(mfVariant: MF.Variant): AST.Variant {
	return {
		match: mfVariant.keys.map((key) => (key.type === "literal" ? key.value : "*")),
		pattern: toPattern(mfVariant.value),
	}
}

function toDeclaration(mfDeclaration: MF.Declaration): AST.Declaration {
	switch (mfDeclaration.type) {
		case "input": {
			return {
				type: mfDeclaration.type,
				name: mfDeclaration.name,
				value: toExpression(mfDeclaration.value),
			}
		}
		default: {
			throw new Error("Unsupported declaration type: " + mfDeclaration.type)
		}
	}
}

function toPattern(mfPattern: MF.Pattern): AST.Pattern {
	return mfPattern.map((element) => {
		if (typeof element == "string")
			return {
				type: "text",
				value: element,
			}

		if (element.type === "expression") return toExpression(element)
		throw new Error("Unknown pattern element type: " + element.type)
	})
}

function toExpression(mfExpression: MF.Expression): AST.Expression {
	const annotation = mfExpression.annotation
	const arg = mfExpression.arg
	if (!arg) throw new Error()
	return {
		type: "expression",
		arg,
		annotation: annotation ? toAnnotation(annotation) : undefined,
	}
}

function toAnnotation(
	mfAnnotation: MF.FunctionAnnotation | MF.UnsupportedAnnotation
): AST.FunctionAnnotation {
	if (mfAnnotation.type !== "function") throw new Error("Junk annotation")
	return {
		type: "function",
		name: mfAnnotation.name,
		options:
			mfAnnotation.options?.map((mfOption) => ({
				name: mfOption.name,
				value: mfOption.value,
			})) || [],
	}
}
