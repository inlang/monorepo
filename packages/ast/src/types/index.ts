export * from "./comments.js";
export * from "./entry.js";
export * from "./expression.js";
export * from "./inlineExpression.js";
export * from "./literal.js";
export * from "./patternElement.js";
export * from "./reference.js";
