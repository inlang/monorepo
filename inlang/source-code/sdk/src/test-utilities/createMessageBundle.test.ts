import { describe, it, expect } from "vitest"
import { createMessageBundle } from "./createMessageBundle.js"
import { MessageBundle } from "../v2/types.js"
import { Value } from "@sinclair/typebox/value"

describe("createMessageBundle", () => {
	it("creates a bundle with no messages", () => {
		const bundle: unknown = createMessageBundle({
			id: "hello_world",
			messages: {},
		})

		expect(Value.Check(MessageBundle, bundle)).toBe(true)

		expect(bundle).toEqual({
			id: "hello_world",
			alias: {},
			messages: [],
		} satisfies MessageBundle)
	})

	it("creates a bundle with messages and an alias", () => {
		const bundle: unknown = createMessageBundle({
			id: "happy_elephant",
			aliases: {
				default: "hello_world",
			},
			messages: {
				en: ".input {$name} {{Hello {$name}}}",
				de: ".input {$name} .match {$name} World {{Hallo Welt}} * {{Hallo {$name}}}",
			},
		})

		expect(Value.Check(MessageBundle, bundle)).toBe(true)
		expect(bundle).toMatchSnapshot()
	})
})
