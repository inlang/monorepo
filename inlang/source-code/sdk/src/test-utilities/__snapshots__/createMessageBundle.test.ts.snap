// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createMessageBundle > creates a bundle with messages and an alias 1`] = `
{
  "alias": {
    "default": "hello_world",
  },
  "id": "happy_elephant",
  "messages": [
    {
      "declarations": [
        {
          "name": "name",
          "type": "input",
          "value": {
            "annotation": undefined,
            "arg": {
              "name": "name",
              "type": "variable",
            },
            "type": "expression",
          },
        },
      ],
      "locale": "en",
      "selectors": [],
      "variants": [
        {
          "match": [],
          "pattern": [
            {
              "type": "text",
              "value": "Hello ",
            },
            {
              "annotation": undefined,
              "arg": {
                "name": "name",
                "type": "variable",
              },
              "type": "expression",
            },
          ],
        },
      ],
    },
    {
      "declarations": [
        {
          "name": "name",
          "type": "input",
          "value": {
            "annotation": undefined,
            "arg": {
              "name": "name",
              "type": "variable",
            },
            "type": "expression",
          },
        },
      ],
      "locale": "de",
      "selectors": [
        {
          "annotation": undefined,
          "arg": {
            "name": "name",
            "type": "variable",
          },
          "type": "expression",
        },
      ],
      "variants": [
        {
          "match": [
            "World",
          ],
          "pattern": [
            {
              "type": "text",
              "value": "Hallo Welt",
            },
          ],
        },
        {
          "match": [
            "*",
          ],
          "pattern": [
            {
              "type": "text",
              "value": "Hallo ",
            },
            {
              "annotation": undefined,
              "arg": {
                "name": "name",
                "type": "variable",
              },
              "type": "expression",
            },
          ],
        },
      ],
    },
  ],
}
`;
